### DISCUSSION TITLE
Blog post hero image as frontmatter

### DISCUSSION THREAD

**Discussion by perpil** [squidfunk/mkdocs-material#5406] [Ideas]
A common layout for blogs is to have a picture above the title and content, see an example [here]( https://h.dhairyashah.dev/)

I'd like to be able to specify an optional hero image as front matter that appears above the title and post header.  When the image is clicked on in excerpts I'd like it to link to the post.  Currently, it doesn't seem like you can have an image above the title without custom themes and when you put an image first it [looks janky](https://speedrun.nobackspacecrew.com/blog/2023/04/10/the-speedrun-mission.html) below the title.

Thought I'd start this discussion to see if there were others with similar thoughts.  Below is how specifying an image might look in front matter.

```
---
draft: false 
date: 2023-04-21
image: path/to/image
categories:
  - Examples
authors: 
  - perpil
---
```

[Edit 4/22] Added screenshot of original because site is now updated with image above title [per suggestion](https://github.com/squidfunk/mkdocs-material/discussions/5406#discussioncomment-5692214).

<img width="1182" alt="image" src="https://user-images.githubusercontent.com/97474956/233792413-9fcff69e-d461-44fc-8dac-2e69e7577cee.png">

**Comment by squidfunk** [comment]
Thanks for suggesting. Have you tried just inserting the image above the title?

``` markdown
---
draft: false 
date: 2023-04-21
categories:
  - Examples
authors: 
  - perpil
---

![](https://placekitten.com/g/800/410)

# Title
```

**Reply by perpil** [reply]
I hadn't thought of that, that does look better.  On the excerpts page I'd prefer the metadata be closer to the title and the image linked to the post, but I'll play around with themes and loop back.

<img width="959" alt="image" src="https://user-images.githubusercontent.com/97474956/233790774-7700764c-dbcf-4852-9066-15f95e8a9035.png">

**Reply by perpil** [reply]
After playing around with this I think it's ok to put the image before the first heading.  I noticed 3 things when doing this.

1. The slug and title is no longer set to the first heading, they are set to the filename.
2. It would be nice to be able to click on any of the metadata, image or excerpt on any index page and be linked to the actual post instead of having to click on "Continue Reading"
3. (Minor) It would be nice if the metadata location on indexes is customizable.  I.e. Top, below the first heading (i.e. the title), bottom.  Right now it is always on the top.

**Reply by squidfunk** [reply]
Thanks for the feedback! We'll definitely consider those points when we work on the blog plugin again ☺️

> The slug and title is no longer set to the first heading, they are set to the filename.

This sounds like a bug, so if you can craft a reproduction, we're happy to look into it.

**Reply by perpil** [reply]
Thinking about this more,setting a hero image per blog entry in frontmatter would really help with the social previews.  Ideally I'd like the specific image of the blog entry to show along with the title for social previews.  I think you can have a static image now, but making it dynamic based on a blog entry would be awesome.

**Reply by squidfunk** [reply]
> Ideally I'd like the specific image of the blog entry to show along with the title for social previews. I think you can have a static image now, but making it dynamic based on a blog entry would be awesome.

I'm not sure I understand this one. You want to image of a blog entry to be used as a background for social cards? That, you can already do, by specifying the path to the image in the front matter, and then pick that up in your custom social card template. There will soon be a video tutorial on social cards that will explain exactly how to pull this off.

**Reply by perpil** [reply]
I believe this is the same idea: #5567 I'll watch the recent [video](https://www.youtube.com/watch?v=4OjnOc6ftJ8) and see if I can figure out how to do it.

**Reply by squidfunk** [reply]
Ah okay, great! Feel free to share. Otherwise, I'll investigate asap.

**Reply by perpil** [reply]
Ok I got something working with a custom social card template and custom front matter.
<img width="1220" alt="image" src="https://github.com/squidfunk/mkdocs-material/assets/97474956/b80b5407-7b82-4ecb-9565-6a4abd5d104f">

It took me a while to figure out how to do tints with the existing documentation.  One thing I couldn't figure out, when I was defining my own variables like this:

```
# Background image
  - &background_image >-
    {{ page.meta.get("background_image", page.background_image) or layout.background_image or "" }}
```

I couldn't figure out how to reference the value of `background_image` so I had to duplicate code
```
- &tint >-
    {%- if page.meta.get("background_image", page.background_image) or layout.background_image -%}
      #000000BB
    {%- else -%}
      #FFFFFF00
    {%- endif -%}
```

Is there a way I can do something like this?

```
- &tint >-
    {%- if background_image -%}
      #000000BB
    {%- else -%}
      {{ background_color }}
    {%- endif -%}
```

**Reply by squidfunk** [reply]
It's currently not possible to define variables in YAML anchors and use them in subsequent anchors or templates. I'm not sure it's a good idea either, though, because then, order of anchors would matter. Thus, a little code duplication is likely more stable and a better approach than defining variables in one template and using them in another. Additionally, YAML anchors cannot be used _inside_ Jinja templates – those are two different languages that are processed in differerent phases.

**Comment by SilentGlasses** [comment]
What @squidfunk suggested works fine, at least for me....  Just need to use a wide enough image instead.

![mkdocs_material_blog_hero](https://github.com/squidfunk/mkdocs-material/assets/22822565/87583373-05d9-4e5f-bef9-8e313b09ddf6)
